libs/ngxtension/create-resource/src/create-resource.ts

```diff
@@ -0,0 +1,285 @@
+
+export function createResource<TValue, TError = string>(
+	source:
+		| Signal<TValue | undefined>
+		| ((source?: TValue | undefined) => Promise<TValue>),
+	fetcher: (
+		source: TValue | undefined,
+	) => Promise<TValue> | (CreateResourceOptions & { initialValue: TValue }),
+	options?: CreateResourceOptions & { initialValue?: undefined },
+): Resource<TValue, TError>;
+
+export function createResource<TValue, TError = string>(
+	source: Signal<TValue> | ((source: TValue) => Promise<TValue>),
+	fetcher: (
+		source: TValue,
+	) => Promise<TValue> | (CreateResourceOptions & { initialValue: TValue }),
+	options: CreateResourceOptions & { initialValue: TValue },
+): Resource<TValue, TError>;
+
+/**
+ *
+ * This function creates a resource that can be used to fetch data from an API or any other source.
+ * It returns an object with the following properties:
+ * - data: a signal that contains the data
+ * - status: a signal that contains the status of the resource
+ * - error: a signal that contains the error if the resource is in an errored state
+ * - loading: a signal that contains a boolean indicating if the resource is loading
+ * - latest: a function that returns the latest value of the resource
+ * - refetch: a function that refetches the resource
+ * - mutate: a function that updates the value of the resource
+ * - destroy: a function that destroys the resource
+ *
+ * @example
+ *
+ * getUser(id: string): Promise<User> {
+ *   return fetch(`/api/users/${id}`).then(res => res.json());
+ * }
+ *
+ * userId = injectQueryParam('userId');
+ *
+ * res = createResource(() => this.getUser(this.userId()));
+ * or
+ * res = createResource(this.userId, this.getUser);
+ *
+ * res.data() // returns undefined
+ * res.loading() // returns true
+ * res.error() // returns undefined
+ * res.latest() // returns undefined
+ *
+ * // After the promise resolves
+ *
+ * res.data() // returns User
+ * res.loading() // returns false
+ * res.error() // returns undefined
+ * res.latest() // returns User
+ *
+ * // After the promise rejects
+ *
+ * res.data() // returns undefined
+ * res.loading() // returns false
+ * res.error() // returns Error
+ * res.latest() // returns undefined
+ *
+ * // After calling refetch
+ *
+ * res.data() // returns undefined
+ * res.loading() // returns true
+ * res.error() // returns undefined
+ * res.latest() // returns User
+ *
+ * // After calling mutate
+ *
+ * res.data() // returns User
+ * res.loading() // returns false
+ * res.error() // returns undefined
+ * res.latest() // returns User
+ *
+ *
+ */
+export function createResource<TValue, TError = string>(...args: unknown[]) {
+	const { source, fetcher, options } = parseArgs<TValue>(args);
+
+	if (fetcher === undefined) {
+		throw new Error('fetcher is required');
+	}
+
+	const value = signal<TValue | undefined>(options.initialValue);
+	const error = signal<TError | undefined>(undefined);
+	const trigger = signal(0);
+	const state = signal<CreateResourceStatus>(
```

Comment:
maybe this can be renamed to 'status' thus matching the Type Parameter itself too?